### PR TITLE
Remove `should_rank_save_checkpoint` property from Trainer

### LINKED ISSUE #9074: [RFC] Deprecate `should_rank_save_checkpoint`
## Proposed refactoring or deprecation

Now that the checkpoint is better consolidated in the training type plugin, we no longer need this property, as this becomes an internal implementation detail of the training type.

Users (via the checkpoint callback) only need to call `trainer.save_checkpoint` and assume the plugin will handle checks surrounding this for them

### Motivation

- Ensure consolidation of saving logic in one place (e.g. all fsspec code for checkpoint paths shoud sit in one place vs. being scattered around the codebase)
- API simpliication: fewer properties exposed on the Trainer

### Pitch

Deprecate this property: https://github.com/PyTorchLightning/pytorch-lightning/blob/538e743f17c7da4624c902f762922e2837661818/pytorch_lightning/trainer/properties.py#L117-L120

Deprecate this property: https://github.com/PyTorchLightning/pytorch-lightning/blob/538e743f17c7da4624c902f762922e2837661818/pytorch_lightning/plugins/training_type/training_type_plugin.py#L313-L316

Move the directory creation from here: https://github.com/PyTorchLightning/pytorch-lightning/blob/538e743f17c7da4624c902f762922e2837661818/pytorch_lightning/callbacks/model_checkpoint.py#L511-L514

Into here: https://github.com/PyTorchLightning/pytorch-lightning/blob/538e743f17c7da4624c902f762922e2837661818/pytorch_lightning/plugins/io/torch_plugin.py#L30-L41

@SeanNaren - this is where we should also have a `rm_checkpoint` on the Checkpoint IO plugin such that we can deprecate this from the model checkpoint: https://github.com/PyTorchLightning/pytorch-lightning/blob/538e743f17c7da4624c902f762922e2837661818/pytorch_lightning/callbacks/model_checkpoint.py#L502-L505

One thing I'm not sure about: because the `should_rank_save_checkpoint` is exposed from the accelerator to the trainer as part of the public trainer API, does this need to go through a full deprecation cycle? Or is a breaking change as part of the plugins API permissible?

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fixes #9074 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -616,7 +616,7 @@ def __resolve_ckpt_dir(self, trainer: "pl.Trainer") -> None:
 
         self.dirpath = ckpt_path
 
-        if not trainer.fast_dev_run and trainer.should_rank_save_checkpoint:
+        if not trainer.fast_dev_run and trainer.training_type_plugin.should_rank_save_checkpoint:
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/model_checkpoint.py:619]: Do we still need this makedirs call here at all? Should this be part of the IO plug-in when saving?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/model_checkpoint.py]: Yeah, I don't think it's required. Updated!

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 1 above)
Comment (jjenniferdai) [reply pytorch_lightning/callbacks/model_checkpoint.py]: i was hoping to remove this too but I think it may require adding makedirs to the TTPs that  do not call IO plugin's save: https://app.circleci.com/pipelines/github/PyTorchLightning/pytorch-lightning/43772/workflows/96d3a810-fd24-4edf-98c0-bee485122df7/jobs/130419,  same as #9615 previous commit tests

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1465,10 +1465,6 @@ def world_size(self) -> int:
         # some training types define a world size
         return getattr(self.accelerator.training_type_plugin, "world_size", 1)
 
-    @property
-    def should_rank_save_checkpoint(self) -> bool:
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:1469]: consider deprecation as it was a public property

**Review by ananthsub — COMMENTED** [review]
@kaushikb11 we can also mark this as deprecated or remove this: https://github.com/PyTorchLightning/pytorch-lightning/blob/0d3325ea206d429f65a52e9d818751a168ec7780/pytorch_lightning/plugins/training_type/training_type_plugin.py#L311-L314
